src/github.com/matrix-org/dendrite/clientapi/writers/register.go

```diff
@@ -0,0 +1,153 @@
+}
+
+// http://matrix.org/speculator/spec/HEAD/client_server/unstable.html#user-interactive-authentication-api
+type userInteractiveResponse struct {
+	Flows     []flow                 `json:"flows"`
+	Completed []types.LoginType      `json:"completed"`
+	Params    map[string]interface{} `json:"params"`
+	Session   string                 `json:"session"`
+}
+
+type flow struct {
+	Stages []types.LoginType `json:"stages"`
+}
+
+func newUserInteractiveResponse(sessionID string, fs []flow) userInteractiveResponse {
+	return userInteractiveResponse{
+		fs, []types.LoginType{}, make(map[string]interface{}), sessionID,
+	}
+}
+
+// http://matrix.org/speculator/spec/HEAD/client_server/unstable.html#post-matrix-client-unstable-register
+type registerResponse struct {
+	UserID      string                       `json:"user_id"`
+	AccessToken string                       `json:"access_token"`
+	HomeServer  gomatrixserverlib.ServerName `json:"home_server"`
+	DeviceID    string                       `json:"device_id"`
+}
+
+// Validate returns an error response if the request fails to validate.
+func (r *registerRequest) Validate() *util.JSONResponse {
+	// https://github.com/matrix-org/synapse/blob/v0.20.0/synapse/rest/client/v2_alpha/register.py#L161
+	if len(r.Password) > 512 {
+		return &util.JSONResponse{
+			Code: 400,
+			JSON: jsonerror.BadJSON("'password' >512 characters"),
+		}
+	} else if len(r.Username) > 512 {
+		return &util.JSONResponse{
+			Code: 400,
+			JSON: jsonerror.BadJSON("'username' >512 characters"),
+		}
+	} else if len(r.Password) > 0 && len(r.Password) < 8 {
+		return &util.JSONResponse{
+			Code: 400,
+			JSON: jsonerror.WeakPassword("password too weak: min 8 chars"),
+		}
+	}
+	return nil
+}
+
+// Register processes a /register request. http://matrix.org/speculator/spec/HEAD/client_server/unstable.html#post-matrix-client-unstable-register
+func Register(req *http.Request, accountDB *storage.AccountDatabase) util.JSONResponse {
+	var r registerRequest
+	resErr := httputil.UnmarshalJSONRequest(req, &r)
+	if resErr != nil {
+		return *resErr
+	}
+	if resErr = r.Validate(); resErr != nil {
+		return *resErr
+	}
+
+	logger := util.GetLogger(req.Context())
+	logger.WithFields(log.Fields{
+		"username":   r.Username,
+		"auth.type":  r.Auth.Type,
+		"session_id": r.Auth.Session,
+	}).Info("Processing registration request")
+
+	// TODO: Shared secret registration (create new user scripts)
+	// TODO: AS API registration
+	// TODO: Enable registration config flag
+	// TODO: Guest account upgrading
+
+	// All registration requests must specify what auth they are using to perform this request
+	if r.Auth.Type == "" {
+		return util.JSONResponse{
+			Code: 401,
+			// TODO: Hard-coded 'dummy' auth for now with a bogus session ID.
+			//       Server admins should be able to change things around (eg enable captcha)
+			JSON: newUserInteractiveResponse("totallyuniquesessionid", []flow{
+				flow{[]types.LoginType{types.LoginTypeDummy}},
```

Comment:
You can drop the typename when initialising structs inside slices so this can be written:

```go
[]flow{{[]types.LoginType{types.LoginTypeDummy}}}
```

instead of

```go
[]flow{flow{[]types.LoginType{types.LoginTypeDummy}}}
```